src/vm/eventpipebuffermanager.cpp

```diff
@@ -0,0 +1,595 @@
+    // This is the second time if this thread did have one or more buffers, but they were full.
+    if(allocNewBuffer && pBuffer != NULL)
+    {
+        allocNewBuffer = !pBuffer->WriteEvent(pEventThread, event, pData, length, pStack);
+    }
+
+    return !allocNewBuffer;
+}
+
+void EventPipeBufferManager::WriteAllBuffersToFile(EventPipeFile *pFile, LARGE_INTEGER stopTimeStamp)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+        PRECONDITION(pFile != NULL);
+    }
+    CONTRACTL_END;
+
+    // TODO: Better version of merge sort.
+    // 1. Iterate through all of the threads, adding each buffer to a temporary list.
+    // 2. While iterating, get the lowest most recent timestamp.  This is the timestamp that we want to process up to.
+    // 3. Process up to the lowest most recent timestamp for the set of buffers.
+    // 4. When we get NULLs from each of the buffers on PopNext(), we're done.
+    // 5. While iterating if PopNext() == NULL && Empty() == NULL, remove the buffer from the list.  It's empty.
+    // 6. While iterating, grab the next lowest most recent timestamp.
+    // 7. Walk through the list again and look for any buffers that have a lower most recent timestamp than the next most recent timestamp.
+    // 8. If we find one, add it to the list and select its most recent timestamp as the lowest.
+    // 9. Process again (go to 3).
+    // 10. Continue until there are no more buffers to process.
+
+    // Niavely walk the circular buffer, writing the event stream in timestamp order.
+    while(true)
+    {
+        EventPipeEventInstance *pOldestInstance = NULL;
+        EventPipeBuffer *pOldestContainingBuffer = NULL;
+        SListElem<EventPipeBufferList*> *pElem = m_pPerThreadBufferList->GetHead();
+        while(pElem != NULL)
+        {
+            EventPipeBufferList *pBufferList = pElem->GetValue();
+
+            // Peek the next event out of the list.
+            EventPipeBuffer *pContainingBuffer = NULL;
+            EventPipeEventInstance *pNext = pBufferList->PeekNextEvent(stopTimeStamp, &pContainingBuffer);
+            if(pNext != NULL)
+            {
+                // If it's the oldest event we've seen, then save it.
+                if((pOldestInstance == NULL) ||
+                   (pOldestInstance->GetTimeStamp().QuadPart > pNext->GetTimeStamp().QuadPart)) 
+                {
+                    pOldestInstance = pNext;
+                    pOldestContainingBuffer = pContainingBuffer;
+                }
+            }
+
+            pElem = m_pPerThreadBufferList->GetNext(pElem);
+        }
+
+        if(pOldestInstance == NULL)
+        {
+            // We're done.  There are no more events.
+            break;
+        }
+
+        // Write the oldest event.
+        pFile->WriteEvent(*pOldestInstance);
+
+        // Pop the event from the buffer.
+        pOldestContainingBuffer->PopNext(stopTimeStamp);
```

Comment:
PopNext is only called once tracing is disabled.  This means that if anyone is writing an event, they are writing it after the stop timestamp.  What we do is this:

- Buffers are always zero'd.
- Read as far into the buffer as we can.  However, if the timestamp is zero or after the stopTimeStamp, ignore it and don't go any further.  Thus, we won't read corrupted data or go further than the write pointer.